pydantic/schema.py

```diff
@@ -395,7 +395,7 @@ def get_flat_models_from_models(models: Sequence[Type['BaseModel']]) -> TypeMode
 
 
 def get_long_model_name(model: TypeModelOrEnum) -> str:
-    return f'{model.__module__}__{model.__name__}'.replace('.', '__')
```

Comment:
I think it's best to keep it this way but just use `__qualname__` instead of `__name__` if possible

Reply:
Yea I think this is a much better approach and do not see why this would not work. Also, I had no idea `__qualname__` existed, so thanks!

A commit has been pushed with this change.